rust/protocol/src/crypto.rs

```diff
@@ -1,156 +1,244 @@
+pub struct DecryptionError(pub &'static str);
 
-#[derive(Debug)]
-pub(crate) enum DecryptionError {
-    /// The key or IV is the wrong length.
-    BadKeyOrIv,
-    /// Either the input is malformed, or the MAC doesn't match on decryption.
-    ///
-    /// These cases should not be distinguished; message corruption can cause either problem.
-    BadCiphertext(&'static str),
-}
+// TODO: Could be nice to have a type-safe library for manipulating units of bytes safely.
+const BITS_PER_BYTE: usize = std::mem::size_of::<u8>() * 8;
 
-fn aes_256_ctr_encrypt(ptext: &[u8], key: &[u8]) -> Result<Vec<u8>, EncryptionError> {
-    let key: [u8; 32] = key.try_into().map_err(|_| EncryptionError::BadKeyOrIv)?;
+/// The length of the key we use for AES encryption in this crate.
+pub const AES_256_KEY_SIZE: usize = 256 / BITS_PER_BYTE;
 
-    let zero_nonce = [0u8; 16];
-    let mut cipher = Aes256Ctr::new(key[..].into(), zero_nonce[..].into());
+/// The size of the generated nonce we use for AES encryption in this crate.
+pub const AES_NONCE_SIZE: usize = 128 / BITS_PER_BYTE;
 
-    let mut ctext = ptext.to_vec();
-    cipher.apply_keystream(&mut ctext);
-    Ok(ctext)
-}
+/// Use AES-256 in CTR mode.
+pub mod ctr {
+    use super::*;
 
-fn aes_256_ctr_decrypt(ctext: &[u8], key: &[u8]) -> Result<Vec<u8>, DecryptionError> {
-    aes_256_ctr_encrypt(ctext, key).map_err(|e| match e {
-        EncryptionError::BadKeyOrIv => DecryptionError::BadKeyOrIv,
-    })
-}
+    pub fn aes_256_ctr_encrypt(ptext: &[u8], key: &[u8; AES_256_KEY_SIZE]) -> Vec<u8> {
+        let zero_nonce = [0u8; AES_NONCE_SIZE];
+        let mut cipher = Aes256Ctr::new(key[..].into(), zero_nonce[..].into());
 
-pub(crate) fn aes_256_cbc_encrypt(
-    ptext: &[u8],
-    key: &[u8],
-    iv: &[u8],
-) -> Result<Vec<u8>, EncryptionError> {
-    match Cbc::<Aes256, Pkcs7>::new_from_slices(key, iv) {
-        Ok(mode) => Ok(mode.encrypt_vec(ptext)),
-        Err(block_modes::InvalidKeyIvLength) => Err(EncryptionError::BadKeyOrIv),
+        let mut ctext = ptext.to_vec();
+        cipher.apply_keystream(&mut ctext);
+        ctext
     }
-}
 
-pub(crate) fn aes_256_cbc_decrypt(
-    ctext: &[u8],
-    key: &[u8],
-    iv: &[u8],
-) -> Result<Vec<u8>, DecryptionError> {
-    if ctext.is_empty() || ctext.len() % 16 != 0 {
-        return Err(DecryptionError::BadCiphertext(
-            "ciphertext length must be a non-zero multiple of 16",
-        ));
+    pub fn aes_256_ctr_decrypt(ctext: &[u8], key: &[u8; AES_256_KEY_SIZE]) -> Vec<u8> {
+        aes_256_ctr_encrypt(ctext, key)
     }
 
-    let mode =
-        Cbc::<Aes256, Pkcs7>::new_from_slices(key, iv).map_err(|_| DecryptionError::BadKeyOrIv)?;
-    mode.decrypt_vec(ctext)
-        .map_err(|_| DecryptionError::BadCiphertext("failed to decrypt"))
-}
+    /// Length in bytes of the [`Hmac`] key used for [`aes_256_ctr_hmac_encrypt`] and
+    /// [`aes_256_ctr_hmac_decrypt`].
+    pub const MAC_KEY_LENGTH: usize = 80 / BITS_PER_BYTE;
```

Comment:
The MAC key isn't 80 bits; only the MAC itself is. The key can in theory be any length and in practice we use a 32-byte key.